man/pione-list-param.1
.TH pione\-list\-param 1 "" "PIONE User Manual"
.SH NAME
.PP
pione\-list\-param \- show a list of parameters in the document or
package.
.SH SYNOPSIS
.PP
\f[B]pione list\-param\f[] [options] location
.SH DESCRIPTION
.PP
\f[B]list\-param\f[] command shows a list of parameters in the document
or package.
The list includes basic parameters only by default, set
\f[C]\-\-advanced\f[] option if you want to show advanced parameters.
.SH OPTIONS
.TP
.B \-a, \-\-advanced
Show advanced parameters.
.RS
.RE
.SH EXAMPLES
.TP
.B pione list\-param example/HelloWorld/HelloWorld.pione
Show basic parameters in the document
\f[C]example/HelloWorld/HelloWorld.pione\f[].
.RS
.RE
.TP
.B pione list\-param example/HelloWorld
Show basic parameters in the package \f[C]example/HelloWorld\f[].
.RS
.RE
.TP
.B pione list\-param \-\-advanced example/HelloWorld
Show basic and advanced parameters in the package
\f[C]example/HelloWorld\f[].
.RS
.RE
.SH REPORTING BUGS
.PP
Report bugs or feature requests to PIONE\[aq]s issue
tracker(https://github.com/pione/pione/issues).
.SH AUTHORS
Keita Yamaguchi.


History.txt
# History

## 0.3.2(2013/11/05)

* Added literate action document.
* Added `update-package-info` subcommand.
* Added `action` subcommand.
* Added `action:list` subcommand
* Added `--timeout SEC` option to pione-client command.
* Changed to stop processing if PIONE methods not found.
* PNML compiler can generate executable PIONE document.
* Fixed many bugs about parameters.
* Fixed many bugs about features.
* Fixed problem of global item `presence_notification_addresses`.
* Renamed and updated `SelectRuleByParam` package.
* Updated many example's packages.

## 0.3.1(2013/10/29)

* Fixed performance bug.
* Enable to generate profile reports of rule application.
* Added PNML compiler.
* Changed to stop processing when action rule happens error.
* Improved pione-clean command and the document, see `man pione-clean`.
* Added `pione` facade command.

## 0.3.0(2013/10/20)

* Improved multi-inheritance package system.
* Added package database.
* Added package cache system.
* Improved language parser and interpreter.
* Refactored many codes.
* Improved test coverage.

## 0.2.2(2013/07/02)

* Enabled to load remote package.
* Added pione-package command.
* Added PIONE archive(".ppg" file).
* Added git package.
* Fixed the problem about pione-client termination.
* Refactored location system.
* Added HTTPS location.
* Added examples.

## 0.2.1(2013/06/15)

* Added basic/advaneced parameter.
* Enabled to run rehearsal test.
* Imporved test coverage.

## 0.2.0(2013/05/31)

* Added initial package system.
* Added pione-val.
* Improved sequence.
* Improved message log framework.
* Added system log framework.
* Changed to unify process log.
* Added endurance test scripts.
* Added embedded FTP server.
* Reduced OS dependencies.
* Added examples of WorkflowPatterns.

## 0.1.4(2013/05/13)

* Sequence based expressions.
* Enable to distribute tasks based on parameters.
* Added many methods.
* Added annotations.
* Added remove and touch operation as data tuple operations.
* Modified update criteria.
* Added empty rule.
* Added new examples.

## 0.1.3(2013/04/25)

* Added new data expression literal "null" for accepting data nonexistence
* Added new data expression property keyword "neglect" and "care" for the update criterion about data mtime
* Changed log location to be in output location
* Changed log format to JSON
* Added pione-log command that generates agent activity log as XES format
* Use locations in tuples
* Enabled to test FTP location handling without setting up FTP server
* Added the feature to create domain information files(".domain_info")

## 0.1.2(2013/04/01)

* Added shorthand notation for setting parameters
* Added ticket concept for sequential processing
* Improved termination sequence
* Refactored parser and transformer

## 0.1.1(2013/03/20)

* Improved performance
* Added the concept of document parameter
* Added --list-params option
* Fixed bugs of features handling
* Improved comment parser

## 0.1.0(2013/02/19)

* First gem release


lib/pione/command/pione-list-param.rb
module Pione
  module Command
    class PioneListParam < BasicCommand
      #
      # basic informations
      #

      command_name "pione list-param"
      command_banner "Show a list of parameters in the document or package."
      PioneCommand.add_subcommand("list-param", self)

      #
      # options
      #

      use_option :color
      use_option :debug

      define_option(:advanced) do |item|
        item.short = "-a"
        item.long = "--advanced"
        item.desc = "show advanced parameters"
        item.value = lambda {|b| b}
      end

      #
      # command lifecycle: setup phase
      #

      setup :target

      # Setup location of the target document.
      def setup_target
        abort("There are no documents or packages.")  if @argv[0].nil?
        @target_location = Location[@argv[0]]
      end

      #
      # command lifecycle: execution phase
      #

      execute :print_list

      # Print a list of parameters in the document.
      def execute_print_list
        # read package
        @package_handler = Package::PackageReader.read(@target_location)
        @env = @package_handler.eval(Lang::Environment.new)

        # print a list
        basic, advanced = Util::PackageParametersList.find(@env, @env.current_package_id)
        if basic.empty? and advanced.empty?
          puts "there are no user parameters in %s" % @env.current_package_id
        else
          if not(basic.empty?)
            print_by_block("Basic Parameters", basic, @env)
          end
          if not(advanced.empty?) and option[:advanced]
            print_by_block("Advanced Parameters", advanced, @env)
          end
        end
      rescue Package::InvalidPackage => e
        abort("Package error: " + e.message)
      rescue Lang::ParserError => e
        abort("Pione syntax error: " + e.message)
      rescue Lang::LangError => e
        abort("Pione language error: %s(%s)" % [e.message, e.class.name])
      end

      private

      #
      # helper methods
      #

      # Print parameters by block.
      def print_by_block(header, params, env)
        unless params.empty?
          puts "%s:" % header
          params.each do |param|
            puts "  %s := %s" % [param.name, param.value.eval(env).textize]
          end
        end
      end
    end
  end
end
